spyder/plugins/externalconsole/plugin.py

```diff
@@ -0,0 +1,335 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""External console plugin."""
+
+# Standard library imports.
+import os
+import sys
+import tempfile
+import platform
+from typing import List
+from distutils.version import LooseVersion
+
+# Third-party imports
+from qtpy.QtWidgets import QMessageBox
+
+# Local imports
+from spyder.utils import programs
+from spyder.utils.misc import get_python_executable
+from spyder.plugins.run.api import (
+    RunContext, run_execute, RunConfiguration, ExtendedRunExecutionParameters,
+    RunResult, RunExecutor)
+from spyder.api.translations import get_translation
+from spyder.api.plugin_registration.decorators import (
+    on_plugin_available, on_plugin_teardown)
+from spyder.api.plugins import Plugins, SpyderPluginV2
+from spyder.plugins.editor.api.run import FileRun, SelectionRun
+from spyder.plugins.externalconsole.api import (
+    ExtConsolePyConfiguration, ExtConsoleShConfiguration)
+from spyder.plugins.externalconsole.widgets.run_conf import (
+    ExternalConsolePyConfiguration, ExternalConsoleShConfiguration)
+
+# Localization
+_ = get_translation('spyder')
+
+
+class ExternalConsole(SpyderPluginV2, RunExecutor):
+    """External console plugin."""
+
+    NAME = 'external_console'
+    REQUIRES = [Plugins.Run]
+    OPTIONAL = [Plugins.Editor]
+    CONF_SECTION = NAME
+    CONF_FILE = False
+
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent, configuration)
+        RunExecutor.__init__(self)
```

Comment:
Seems like all the plugins that will interact with Run as executors will then need to overwritte their `__init__`, right? Maybe could be worthy to think in a way to define plugins that will use the Run API without each plugin having to overwrite their `__init__` in their definition (not sure if possible but leaving the comment to think about it).

Reply:
It depends if a `RunExecutor` will be restricted to be a graphical plugin or not, in that case, then it would inherit from `SpyderDockablePlugin`. Else, it would inherit from `SpyderPluginV2`, in which case then we would have the so-called [inheritance diamond problem](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem). Since the constructor for `RunExecutor` does not comply with the same signature as `SpyderDockablePlugin`, then the Python MRO will not call the constructor via `super()`, which means that the constructor needs to be called manually

Reply:
🤔 could be possible for the plugin to instead of inheriting the `RunExecutor` class to create an subclass/instance of it and register that? Something like what is done for preferences config pages?

Reply:
Let us try setting the same constructor arguments to see if the constructor gets called

Reply:
The solution was to make sure that `RunExecutor` inherits from `QObject`, this ensures that all the class hierarchy share the same base class

Reply:
Nice! :+1: